### PR TITLE
[Feature] Adds for/against type + debate descriptions

### PR DESCRIPTION
**Pull request by Sqyphen**
## Description

Adds:
- small issue fixes
- minor css amendments
- adds basic debate type
- adds for/against debate type
- adds descriptions for each type of debate
- adds voting mechanic
- turns voting and comments into components for easier reuse

## Implementation

- New debate type can be select from the 'Create' section in the admin
- Descriptions for each debate can be controlled in the _data/debates.json_ file. Updates made here will automatically be pulled through on next page refresh. Plaintext support only right now.
- Comments and votes have been componentised(?) and have their own partials, scss and js files now
- Voting can be enabled/disabled for a debate
- Users can vote and change their vote until the voting period ends
- Closing the debate and the voting displays the voting results

## Screenshot

n/a

## .env changes

n/a

## Additional requirements

n/a

### CODE REVIEW

File: helpers/comments.js [hunk 1]
```diff
@@ -1,21 +1,23 @@
+const Utils = require('../helpers/utils');
+
 function getCommentReplies(comment, commentsReplies) {
   const replies = [];
 
-  commentsReplies.map(c => {
+  commentsReplies.forEach(c => {
```
Comment (Lily2point0) [inline helpers/comments.js]: Can we avoid single letter variables, please?

Comment (Sqyphen) [reply helpers/comments.js]: Yes. Fixed, thanks

File: routes/router.js [hunk 2]
```diff
@@ -28,12 +30,13 @@ router.get('/', async (req, res) => {
 
     res.render('list', {
       pageTitle: 'FT Debates',
-      pageSubtitle: 'Welcome to FT debates, here\'s a list of all available debates and a bit more blurb on how to take part',
+      pageSubtitle:
```
Comment (Lily2point0) [inline routes/router.js]: Looks like a formatting issue here.

File: routes/router.js [hunk 3]
```diff
@@ -28,12 +30,13 @@ router.get('/', async (req, res) => {
...
+        "Welcome to FT debates, here's a list of all available debates and a bit more blurb on how to take part",
       pageType: 'home',
       debateList,
       user: {
         username,
-      }
+      },
```
Comment (Lily2point0) [inline routes/router.js]: trailing comma?

File: static/js/admin/main.js [hunk 4]
```diff
@@ -1,209 +1,237 @@
...
-    var debateStatus = document.getElementsByName('debateStatus')[0].value;
-    var votingStatus = document.getElementsByName('votingStatus')[0].value;
+function updateDescription(e) {
+  var typeValue = e ? e.target.value : selectDebateType.value;
+  var debateDescriptions = document.getElementsByClassName('debateDescription');
 
-    if (isAlphaNumericWithCharacters(title)) {
-        errors.title = 'Title must be alphanumeric or allowed chars (,_\"-\')';
-    }
+  Array.from(debateDescriptions).forEach(el => {
```
Comment (Lily2point0) [inline static/js/admin/main.js]: Front-end JS won't always support arrow notation. Replace with `function()`

File: static/js/admin/main.js [hunk 5]
```diff
@@ -1,209 +1,237 @@
...
-            'Content-Type': 'application/json'
-        },
-        body: JSON.stringify(data)
-    })
+  //TODO: fetch compatibility??
+  return fetch(url, {
+    method: 'POST',
+    headers: {
+      Accept: 'application/json',
+      'Content-Type': 'application/json',
```
Comment (Lily2point0) [inline static/js/admin/main.js]: Trailing commas

File: routes/router.js (hunk 2 above)
Comment (Sqyphen) [reply routes/router.js]: Ah, that's Prettier - making sure lines do not exceed a certain character length

File: routes/router.js (hunk 3 above)
Comment (Sqyphen) [reply routes/router.js]: Trailing commas are on as part of Prettier

File: static/js/components/comments.js [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+  );
+  commentReplyNotification.classList.add('hidden');
+  replyTo.removeAttribute('value');
+}
+
+function addReplyEventListeners() {
+  const replyLinks = document.querySelectorAll('.reply');
+  const commentReplyRemove = document.querySelector('.comment-reply-remove');
+
+  Array.from(replyLinks).forEach(element =>
```
Comment (Lily2point0) [inline static/js/components/comments.js]: Arrow function here too

File: views/partials/components/heading/topper.hbs [hunk 7]
```diff
@@ -0,0 +1,9 @@
+<div class="debate-topper-container">
+	<div class="o-grid-container">
+		<div class="o-grid-row debate-topper">
+			<p class="debate-type">{{debateType}}</p>
```
Comment (Lily2point0) [inline views/partials/components/heading/topper.hbs]: DebateType appears empty on ForAgainst debates.

Comment (Lily2point0) [reply views/partials/components/heading/topper.hbs]: Looks like the way the data is passed in from the server.

File: static/js/admin/main.js (hunk 5 above)
Comment (Sqyphen) [reply static/js/admin/main.js]: Will correct in a future formatting PR

File: routes/router.js (hunk 3 above)
Comment (Sqyphen) [reply routes/router.js]: Will correct in a future formatting PR

File: static/js/admin/main.js (hunk 4 above)
Comment (Sqyphen) [reply static/js/admin/main.js]: Corrected

File: views/debates/interview.hbs [hunk 8]
```diff
@@ -0,0 +1,17 @@
+<!DOCTYPE html>
```
Comment (Lily2point0) [inline views/debates/interview.hbs]: admin landing pages seem to be missing the HTML declaration and opening tag at the top of the page. were they part of the head partial before?

File: views/partials/components/heading/topper.hbs (hunk 7 above)
Comment (Sqyphen) [reply views/partials/components/heading/topper.hbs]: Fixed

File: views/debates/foragainst.hbs [hunk 9]
```diff
@@ -0,0 +1,64 @@
...
+            <div class="comments-for" data-o-grid-colspan="6">
+                <h2>For</h2>
+                {{#if commentsFor}}
+                {{> components/comments/list class='comments_list' comments=commentsFor debateId=debateId repliesOpen=false }}
+                {{/if}}
+            </div>
+        </div>
+    </main>
+
+    {{#if debateOpen}}
```
Comment (Lily2point0) [inline views/debates/foragainst.hbs]: The comment box is obscuring the last comment on the page, cannot scroll past.
Screenshot 2019-05-09 at 14 11 08

File: views/debates/interview.hbs (hunk 8 above)
Comment (Sqyphen) [reply views/debates/interview.hbs]: Wow, nice spot.
Not sure where they were before but I'm adding them to the head for admin and user pages

File: views/debates/basic.hbs [hunk 10]
```diff
@@ -0,0 +1,40 @@
...
+        <div class="o-grid-row">
+            <div class="comments-against" data-o-grid-colspan="6">
+                <h2>Against</h2>
+                {{#if commentsAgainst}}
+                {{>  components/comments/list class='comments_list' comments=commentsAgainst debateId=debateId  repliesOpen=true }}
+                {{/if}}
+            </div>
+            <div class="comments-for" data-o-grid-colspan="6">
+                <h2>For</h2>
+                {{#if commentsFor}}
```
Comment (Lily2point0) [inline views/debates/basic.hbs]: Vertical margin between reply boxes to a top comment is a bit off. (We might want to make the link more obvious for this, but can work on at a later date, just making a note of it.
Screenshot 2019-05-09 at 14 14 07

File: views/debates/foragainst.hbs (hunk 9 above)
Comment (Sqyphen) [reply views/debates/foragainst.hbs]: eeep, fixed

File: static/js/components/comments.js (hunk 6 above)
Comment (Sqyphen) [reply static/js/components/comments.js]: Yes. Fixed now

File: views/debates/basic.hbs (hunk 10 above)
Comment (Sqyphen) [reply views/debates/basic.hbs]: Okay - have fix the weird margin issue but agree that this could be clearer

**Pull request merged**